WcaOnRails/app/models/competition.rb

```diff
@@ -1737,29 +1738,33 @@ def update_persons_wcif!(wcif_persons, current_user)
         registration.update!(roles: roles)
       end
       if registration && wcif_person["assignments"]
-        registration.assignments = wcif_person["assignments"].map do |assignment_wcif|
-          existing_assignment = registration.assignments.find do |assignment|
-            assignment.wcif_equal?(assignment_wcif)
+        wcif_person["assignments"].map do |assignment_wcif|
+          schedule_activity = competition_activities.find do |competition_activity|
+            competition_activity.wcif_id == assignment_wcif["activityId"]
           end
-          if existing_assignment
-            existing_assignment
+          unless schedule_activity
+            raise WcaExceptions::BadApiParameter.new("Cannot create assignment for non-existent activity with id #{assignment_wcif["activityId"]}")
+          end
+          # We need to be very careful about how we build the assignment:
+          # providing just the registration_id or the schedule_activity_id would
+          # actually trigger a select for each validation.
+          a = registration.assignments.build(
+            schedule_activity: schedule_activity,
+            station_number: assignment_wcif["stationNumber"],
+            assignment_code: assignment_wcif["assignmentCode"],
+          )
+          if a.valid?
+            new_assignments << a.attributes
           else
-            schedule_activity = competition_activities.find do |competition_activity|
-              competition_activity.wcif_id == assignment_wcif["activityId"]
-            end
-            unless schedule_activity
-              raise WcaExceptions::BadApiParameter.new("Cannot create assignment for non-existent activity with id #{assignment_wcif["activityId"]}")
-            end
-            registration.assignments.build(
-              schedule_activity_id: schedule_activity.id,
-              station_number: assignment_wcif["stationNumber"],
-              assignment_code: assignment_wcif["assignmentCode"],
-            )
+            raise WcaExceptions::BadApiParameter.new("Invalid assignment: #{a.errors.map(&:full_message)} for #{assignment_wcif}")
           end
         end
-        registration.save!
       end
     end
+    Assignment.where(registration_id: registrations.ids).delete_all
```

Comment:
I am not too wild about flat out deleting all existing assignments. This leads to conflicts when two different WCIF tools write assignments and sync them back to the website, where the first can create an assignment that is later deleted and re-created with different values by a second tool

Reply:
I'm not too wild either, but rather because it creates unnecessary read/write to the db; as far as the WCIF and external tools are concerned there is absolutely no changes: the WCIF values (`activityId`/`stationNumber`/`assignmentCode`) stay the same, and therefore external tools would see the "same" assignment.
What does change is our internal database id associated with that assignment.

I'll try to see if I can come up with something better.

Reply:
I just pushed an implementation which tries to re-use existing assignment, but to be honest it doesn't change much in practice: either with the current implementation (in master) or the one I'm suggesting here we rely on the whole WCIF attributes to figure out if assignments are equal.
It's the only solution because assignments don't have ids, but it means that if we change one single attribute (eg: the station number) we don't find the existing assignment and we end up deleting the old one and creating a new one.

I've still pushed the new version because it avoids deleting everything if nothing changed.